labs/b4.md

```diff
@@ -0,0 +1,167 @@
+
+**This, and all other parts of the lab, should be done on your DigitalOcean VMs, as you cannot install packages on OCF machines.** 
+Please note that you can only connect to your VM from inside the Berkeley network, so you will have to either be on campus wifi, or you will have to SSH 
+through `ssh.ocf.berkeley.edu` (or a similar on-campus host) to access your VM. 
+You should be able to connect to it at `<OCF username>.decal.xcf.sh` by running `ssh <username>@<username>.decal.xcf.sh`.
+
+Simply run:
+
+`sudo apt install gcc`
+
+Now check if GCC is installed by running the following to check GCC's version:
+
+`gcc --v`
+
+Now install the package `ruby-dev` for the next parts of the lab.
+
+## But what about software that isn’t in the repositories/package lists?
+
+This is a very common and frustrating thing to occur. Just when you want to install a package, it appears that it is not listed in any of the repositories that are available. What can you do? There are three ways in going about this. 
+
+#### Warning:
+**Installing software not from a repository carries the same risks as installing software from a .exe or .msi on Windows.** Linux is popularly considered more secure than Windows or MacOS because of its use of repositories--but a malicious package installed manually can pwn your system as thoroughly as a malware executable can on other OSes. **Use common sense! Only manually install from sources and companies you trust.**
+
+*(For users of Ubuntu: PPAs, or third-party repositories, carry the same risks of running malicious third-party code.)*
+
+If the developers provide a package compatible with your distribution (.deb for Debian-based distros), you can download that package and install it using:
+
+`apt install ./package.deb`
+
+Another way is to find a generic binary package from the developer. This comes in the form of shady shell scripts, or a binary tarball in which you can just extract it and run it, or an appimage, which is a special type of executable that brings its own dependencies.
+
+The last way is to compile your own software from scratch. What does that mean? Free and open-source software usually has their source code somewhere publicly available (GitHub, GitLab, their website). If you fetch their source code, it won’t magically run out of the box. The source code is like the recipe, while the software itself is like food. A package is like a food that is put into a box but we won’t covering the details of making a package yourself (there are tools that do that for you and it varies from platform to platform).
+
+## So how do I compile?
+
+Compiling software on Linux is a mixed bag. Sometimes, all the dependencies are installed on your computer and there is no fiddling around. Sometimes, the dependencies don’t even exist pre-compiled for your distribution so you have to compile those yourself in order to compile what’s at hand. But all of these steps all have one component in common: the makefile.
```

Comment:
Make is hardly the only build system used, we probably shouldn't say "all these steps have one component in common" (also all used twice in that sentence)